### ISSUE TITLE
Partial-izing the template does not behave as expected

### ISSUE THREAD

**Issue by ktalebian** [Microsoft/TypeScript#15169]
Here is an interface that takes in two templates. the returned function should be a partial of one of the templates (this is a take on the React-Redux):

```ts
export type IState<TState, TOwnProps> = {
    connect: (mapStateToProps: MapStateToProps<TState, Partial<TOwnProps>>) => any;
}
export type MapStateToProps<TState, TProps> = (state: TState) => TProps;
```
Now, here is the example of a state that would want to use this:

```ts
interface IAppState {
    services: string[]
}
    
interface IAppProps {
    services: string[];
    count: number;
}
    
const MyState: IState<IAppState, IAppProps>;
MyState.connect(
    (state) => {
        return {
            services: state.services,
            foo: null
        }
    }
);
```

`MyState.connect` should expect one function (which it does). The argument should be of type `IAppState` (which it is). But the return is suppose to be a `Partial<IAppProps>`. However, as you can see I am returning `foo: null` too and the service is not complaining about this. 

What am I doing wrong here?

**Comment by mhegazy** [comment]
This is not specific to map types, or `Partial`. the excess property checks (that should catch `foo` as an extra property) do not run on return positions of contextually typed functions. https://github.com/Microsoft/TypeScript/issues/12632 tracks that.
